std/experimental/allocator/building_blocks/allocator_list.d

```diff
@@ -0,0 +1,614 @@
+{
+    import std.traits : hasMember;
+    import std.conv : emplace;
+    import std.algorithm : min, move;
+    import std.experimental.allocator.building_blocks.stats_collector
+        : StatsCollector, Options;
+
+    private enum ouroboros = is(BookkeepingAllocator == NullAllocator);
+
+    /**
+    Alias for `typeof(Factory()(1))`, i.e. the type of the individual
+    allocators.
+    */
+    alias Allocator = typeof(Factory.init(1));
+    // Allocator used internally
+    private alias SAllocator = StatsCollector!(Allocator, Options.bytesUsed);
+
+    private static struct Node
+    {
+        // Allocator in this node
+        SAllocator a;
+        Node* next;
+
+        @disable this(this);
+
+        // Is this node unused?
+        void setUnused() { next = &this; }
+        bool unused() const { return next == &this; }
+
+        // Just forward everything to the allocator
+        alias a this;
+    }
+
+    /**
+    If $(D BookkeepingAllocator) is not $(D NullAllocator), $(D bkalloc) is
+    defined and accessible.
+    */
+
+    // State is stored in an array, but it has a list threaded through it by
+    // means of "nextIdx".
+
+    // state {
+    static if (!ouroboros)
+    {
+        static if (stateSize!BookkeepingAllocator) BookkeepingAllocator bkalloc;
+        else alias bkalloc = BookkeepingAllocator.instance;
+    }
+    static if (stateSize!Factory)
+    {
+        Factory factory;
+    }
+    private Node[] allocators;
+    private Node* root;
+    // }
+
+    static if (stateSize!Factory)
+    {
+        private auto make(size_t n) { return factory(n); }
+    }
+    else
+    {
+        private auto make(size_t n) { Factory f; return f(n); }
+    }
+
+    /**
+    Constructs an `AllocatorList` given a factory object. This constructor is
+    defined only if `Factory` has state.
+    */
+    static if (stateSize!Factory)
+    this(ref Factory plant)
+    {
+        factory = plant;
+    }
+    /// Ditto
+    static if (stateSize!Factory)
+    this(Factory plant)
+    {
+        factory = plant;
+    }
+
+    static if (hasMember!(Allocator, "deallocateAll")
+        && hasMember!(Allocator, "owns"))
+    ~this()
+    {
+        deallocateAll;
+    }
+
+    /**
+    The alignment offered.
+    */
+    enum uint alignment = Allocator.alignment;
+
+    /**
+    Allocate a block of size $(D s). First tries to allocate from the existing
+    list of already-created allocators. If neither can satisfy the request,
+    creates a new allocator by calling $(D make(s)) and delegates the request
+    to it. However, if the allocation fresh off a newly created allocator
+    fails, subsequent calls to $(D allocate) will not cause more calls to $(D
+    make).
+    */
+    void[] allocate(size_t s)
+    {
+        for (auto p = &root, n = *p; n; p = &n.next, n = *p)
+        {
+            auto result = n.allocate(s);
+            if (result.length != s) continue;
+            assert(owns(result) == Ternary.yes);
```

Comment:
Disregard that, it was a unit test for an `FreeList!(AllocatorList!(n => Region!Mallocator(blockSize)), ns)` that I had in my code somewhere that failed.